src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -159,31 +158,45 @@ public bool CanRetry
 
         public DateTimeOffset CreationTime { get; } = DateTimeOffset.UtcNow;
 
+        private ReadOnlySpan<byte> RemainingBuffer => new ReadOnlySpan<byte>(_readBuffer, _readOffset, _readLength - _readOffset);
+
+        private void ConsumeRemainingBuffer(int bytes)
+        {
+            Debug.Assert(_readOffset <= _readLength - bytes, $"{_readOffset} + {bytes} > {_readLength}");
+            _readOffset += bytes;
+        }
+
         private async Task WriteHeadersAsync(HttpHeaders headers, string cookiesFromContainer)
         {
-            foreach (KeyValuePair<string, IEnumerable<string>> header in headers)
+            foreach (KeyValuePair<HeaderDescriptor, string[]> header in headers.GetHeaderDescriptorsAndValues())
             {
-                await WriteAsciiStringAsync(header.Key).ConfigureAwait(false);
-                await WriteTwoBytesAsync((byte)':', (byte)' ').ConfigureAwait(false);
+                if (header.Key.KnownHeader != null)
+                {
+                    await WriteBytesAsync(header.Key.KnownHeader.AsciiBytesWithColonSpace).ConfigureAwait(false);
+                }
+                else
+                {
+                    await WriteAsciiStringAsync(header.Key.Name).ConfigureAwait(false);
+                    await WriteTwoBytesAsync((byte)':', (byte)' ').ConfigureAwait(false);
+                }
 
-                var values = (string[])header.Value; // typed as IEnumerable<string>, but always a string[]
-                Debug.Assert(values.Length > 0, "No values for header??");
-                if (values.Length > 0)
+                Debug.Assert(header.Value.Length > 0, "No values for header??");
+                if (header.Value.Length > 0)
                 {
-                    await WriteStringAsync(values[0]).ConfigureAwait(false);
+                    await WriteStringAsync(header.Value[0]).ConfigureAwait(false);
 
-                    if (cookiesFromContainer != null && header.Key == HttpKnownHeaderNames.Cookie)
+                    if (cookiesFromContainer != null && ReferenceEquals(header.Key.KnownHeader, KnownHeaders.Cookie))
```

Comment:
Is the ReferenceEquals necessary?

Reply:
No, it was just more clear to me, as its semantics can't be changed whereas =='s can.  I can change it back, though.